File: Data Structures & Algorithms/binary-search/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int first=0,last=nums.size()-1;
        while(first<=last)// why =?---> because we need the last element chack
        {
           int middle=(first+last)/2;
           if(nums[middle]>target){ last=middle-1;}// also some major edge cases
           else if (nums[middle]<target){ first=middle+1;}
else {return middle;}
        }
       return -1; 
    }
};

File: Data Structures & Algorithms/sqrtx/submission-1.py
class Solution:
    def mySqrt(self, x: int) -> int:
        l=0
        r=x
        while l<=r:
            m=(l+r)//2
            s=m*m
            if s== x:
                return m
            elif s>x:
                r=m-1
            else:
                l=m+1
        return r

        